formatters/critical-network-chains.js

```diff
@@ -0,0 +1,220 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const url = require('url');
+const path = require('path');
+const fs = require('fs');
+const Formatter = require('./formatter');
+const html = fs.readFileSync(path.join(__dirname, 'partials/critical-network-chains.html'), 'utf8');
+
+class CriticalNetworkChains extends Formatter {
+
+  /**
+   * gets the formatter for the CLI Printer and the HTML report.
+   */
+  static getFormatter(type) {
+    switch (type) {
+      case 'pretty':
+        return function(info) {
+          const longestChain = CriticalNetworkChains._getLongestChainLength(info);
+          const longestDuration = CriticalNetworkChains._getLongestChainDuration(info);
+          const urlTree = CriticalNetworkChains._createURLTreeOutput(info);
+
+          const output = `    - Longest request chain (shorter is better): ${longestChain}\n` +
+          `    - Longest chain duration (shorter is better): ${longestDuration.toFixed(2)}ms\n` +
+          '    - Chains\n' +
+              '      ' + urlTree.replace(/\n/g, '\n      ') + '\n';
+          return output;
+        };
+
+      case 'html':
+        // Returns a handlebars string to be used by the Report.
+        return html;
+
+      default:
+        throw new Error('Unknown formatter type');
+    }
+  }
+
+  static _getLongestChainLength(info) {
+    return info.reduce((total, item) => {
+      if (item.totalRequests > total) {
+        return item.totalRequests;
+      }
+
+      return total;
+    }, 0);
+  }
+
+  static _getLongestChainDuration(info) {
+    return info.reduce((total, item) => {
+      if (item.totalTimeBetweenBeginAndEnd > total) {
+        return item.totalTimeBetweenBeginAndEnd;
+      }
+
+      return total;
+    }, 0);
+  }
+
+  /**
+   * Refactors the gatherer's request information into a tree.
+   */
+  static _createURLTree(info) {
+    return info.reduce((tree, item) => {
+      let node = tree;
+      item.urls.forEach((itemURL, index, arr) => {
+        if (!node[itemURL]) {
+          const isLastChild = (index === arr.length - 1);
+          node[itemURL] = isLastChild ? item.totalTimeBetweenBeginAndEnd : {};
```

Comment:
It feels a little silly that we create the whole tree in the gatherer, then extract the chains out of it and throw the tree away, and then recreate the tree again here. We should be able to pass through the whole tree from the gatherer. We can do this refactor in a separate PR though.